idaes/apps/grid_integration/examples/thermal_generator_storage.py

```diff
@@ -0,0 +1,1069 @@
+
+def _relax_pbc_constriant(m):
+
+    n_timesteps = len(m.HOUR)
+
+    for j in m.UNITS:
+        for t in m.HOUR:
+            if t == n_timesteps - 1:
+                m.pbc_slack[j,t].unfix()
+                m.pbc_slack_positive[j,t].unfix()
+                m.pbc_slack_negative[j,t].unfix()
+
+    return
+
+def _enforce_pbc_constriant(m):
+
+    for j in m.UNITS:
+        for t in m.HOUR:
+            m.pbc_slack[j,t].fix(0)
+            m.pbc_slack_positive[j,t].fix(0)
+            m.pbc_slack_negative[j,t].fix(0)
+
+    return
+
+def _unfix_slack_vars(m):
+
+    # unfix pbc slack vars
+    n_timesteps = len(m.HOUR)
+    for j in m.UNITS:
+        for t in m.HOUR:
+            if t == n_timesteps - 1:
+                m.pbc_slack[j,t].unfix()
+                m.pbc_slack_positive[j,t].unfix()
+                m.pbc_slack_negative[j,t].unfix()
+
+    # slack variables for power dispatch
+    m.slack_var_power.unfix()
+
+    return
+
+def _fix_slack_vars(m):
+
+    m.pbc_slack.fix(0)
+    m.pbc_slack_positive.fix(0)
+    m.pbc_slack_negative.fix(0)
+
+    m.slack_var_power.fix(0)
+
+    return
+
+def _get_min_slack(m):
+
+    # get the min slack values for pbc
+    for j in m.UNITS:
+        total_slack = 0
+        for h in m.HOUR:
+            total_slack += pyo.value(m.pbc_slack_positive[j,h] + m.pbc_slack_negative[j,h])
+        m.pbc_slack_bound[j] = total_slack
+
+    for j in m.UNITS:
+        for h in m.HOUR:
+            m.power_slack_bound[j,h] = pyo.value(m.slack_var_power[j,h])
+
+    return
+
+def activate_relaxed_mode(m):
+
+    print('')
+    print('Activating relaxing mode...')
+
+    _unfix_slack_vars(m)
+
+    # deactivate original constraint and activate min slack constraint
+    m.abs_dev.deactivate()
+    m.min_slack.activate()
+
+    # deactivate original constraints
+    m.fix_power_output_con.deactivate()
+    m.SOC_endtime_con.deactivate()
+
+    # activate the relaxed constraints
+    m.power_slack_con1.activate()
+    m.power_slack_con2.activate()
+    m.pbc_slack_con.activate()
+    m.SOC_endtime_slack_con.activate()
+
+    return
+
+def activate_bounded_slack_mode(m):
+
+    _get_min_slack(m)
+
+    # solve the original problem with bounded slack
+    m.bound_pbc_slack_con.activate()
+    m.bound_power_slack_con.activate()
+
+    m.min_slack.deactivate()
+    m.abs_dev.activate()
+
+    return
+
+def deactivate_relaxed_mode(m):
+
+    print('')
+    print('Deactivating relaxing mode..')
+
+    _fix_slack_vars(m)
+
+    m.min_slack.deactivate()
+    m.abs_dev.activate()
+
+    # deactivate slack constraints
+    m.bound_pbc_slack_con.deactivate()
+    m.bound_power_slack_con.deactivate()
+    m.power_slack_con1.deactivate()
+    m.power_slack_con2.deactivate()
+    m.pbc_slack_con.deactivate()
+    m.SOC_endtime_slack_con.deactivate()
+
+    # activate original cosntraints
+    m.fix_power_output_con.activate()
+    m.SOC_endtime_con.activate()
+
+    return
+
+def solve_relaxed_model(m,solver):
+    '''
```

Comment:
Should this be part of the generalized workflow? My gut says yes, as a PBC will be a standard feature in most of the IES models.